snap/quota/resources.go

```diff
@@ -0,0 +1,123 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+* Copyright (C) 2021 Canonical Ltd
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License version 3 as
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+ */
+
+package quota
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+// ResourceMemory is the memory limit for the quota group being controlled,
+// either the initial limit the group is created with for the "create"
+// action, or if non-zero for the "update" the memory limit, then the new
+// value to be set.
+type ResourceMemory struct {
+	// Lets not set omitempty here, as we want to be able to see the 0 value in
+	// case the memory struct is provided. Remember that we have omitempty set on
+	// the memory value in QuotaResources.
+	Limit quantity.Size `json:"limit"`
+}
+
+// Resources is built up of multiple quota limits. Each quota limit is a pointer
+// value to indicate that their presence may be optional, and because we want to detect
+// whenever someone changes a limit to '0' explicitly.
+type Resources struct {
+	Memory *ResourceMemory `json:"memory,omitempty"`
+}
+
+func (qr *Resources) validateMemoryQuota() error {
+	// make sure the memory limit is not zero
+	if qr.Memory.Limit == 0 {
+		return fmt.Errorf("cannot create quota group with no memory limit set")
+	}
+
+	// make sure the memory limit is at least 4K, that is the minimum size
+	// to allow nesting, otherwise groups with less than 4K will trigger the
+	// oom killer to be invoked when a new group is added as a sub-group to the
+	// larger group.
+	if qr.Memory.Limit <= 4*quantity.SizeKiB {
+		return fmt.Errorf("memory limit %d is too small: size must be larger than 4KB", qr.Memory.Limit)
+	}
+	return nil
+}
+
+// Validate performs validation of the provided quota resources for a group. Its intended
+// use is before creating a group.
+func (qr *Resources) Validate() error {
```

Comment:
hm wonder if the name should be changed to something like ValidateForCreate() to indicate that this is not a generic validation but rather just for the code path that creates quota groups. Then the other one could be named ValidateForChange() ? @anonymouse64 wdyt?

Reply:
this sounds like a good idea

Reply:
This still needs addressing

Reply:
Ok, so I was going to change this, but looking around the code it's not entirely correct. The Validate() is actually called as a generic validation (during group.validate()) as well as initial validation, so the fix here is to make sure the error message is a bit more generic.